iree/hal/local/task_device.h

```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_HAL_LOCAL_TASK_DEVICE_H_
+#define IREE_HAL_LOCAL_TASK_DEVICE_H_
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/local/executable_loader.h"
+#include "iree/task/executor.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// Parameters configuring an iree_hal_task_device_t.
+// Must be initialized with iree_hal_task_device_params_initialize prior to use.
+typedef struct {
+  // Number of queues exposed on the device.
+  // Each queue acts as a separate synchronization scope where all work executes
+  // concurrently unless prohibited by semaphores.
+  iree_host_size_t queue_count;
+
+  // Total size of each block in the device shared block pool.
+  // Larger sizes will lower overhead and ensure the heap isn't hit for
+  // transient allocations while also increasing memory consumption.
+  iree_host_size_t arena_block_size;
+} iree_hal_task_device_params_t;
+
+// Initializes |out_params| to default values.
+void iree_hal_task_device_params_initialize(
+    iree_hal_task_device_params_t* out_params);
+
+// Creates a new iree/task/-based local CPU device that uses |executor| for
+// scheduling tasks. |loaders| is the set of executable loaders that are
```

Comment:
Hmm.. the name "task device" isn't very inspiring. My main gripe with it is that it assumes knowledge about iree/task/ as a system large enough to differentiate it from other 'device's like "vulkan device". I'd almost prefer "host device", "local device", or "cpu device". WDYT?

(same with "task driver" and other objects)

Reply:
That's the point :P This is a device for local execution _implemented using the task system_. It's also possible to implement a local device *without* the task system - like for a single threaded system almost all of the task system is overhead (besides queue ordering) and a much simpler implementation can be added. The split in those files is because local_* files are reusable while the task_* files are specific to the implementation using the task system. It's precisely equivalent to a device that uses vulkan for execution or metal for execution: the task system is the thing providing the execution.

The name "task" won't be exposed to users - it'll just be "local" and which implementation (task or other) will be controlled by build flags or runtime options. That's part of the dylib/vmla reworking: both of those just become build flags and runtime options for which executable loaders to support (including allowing out-of-tree loaders).